### PR TITLE
Client frontend

### PR DESCRIPTION
**Pull request by ChihChengLiang**

### CODE REVIEW

File: contracts/client/ClientFrontend.sol [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+            newRoot,
+            _tx.toIndex,
+            _tx.amount,
+            tokenType,
+            to
+        );
+        return (newRoot, senderState, receiverState, result);
+    }
+
+    function applyTransferTx(
```
Comment (ChihChengLiang) [inline contracts/client/ClientFrontend.sol:78]: @vaibhavchellani  Does the go client require state root to be exposed in this API? [This line](https://github.com/thehubbleproject/reddit-commander/blob/ad827761d50f9b7989dc22286f18163ec858e204/core/tx.go#L92) suggests it doesn't, but not sure if I'm checking the right line

Comment (vaibhavchellani) [reply contracts/client/ClientFrontend.sol:78]: Ideally I would want it to return the newState after apply, I am not sure about the purpose right now but I can see a usecase where a user would want `fast-sync`, and during that we can just apply transactions without processTx and just validate state updates via stateRoot check. 

But I think we can skip for now and add that later.

File: contracts/client/Offchain.sol [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    struct Transfer {
+        uint256 txType;
+        uint256 fromIndex;
+        uint256 toIndex;
+        uint256 amount;
+        uint256 fee;
+        uint256 nonce;
+    }
+
+    function decodeTransfer(bytes memory txBytes)
```
Comment (ChihChengLiang) [inline contracts/client/Offchain.sol:16]: Should check if the amount and fee passed here are compressible. Will think about it more in the next pr

File: contracts/client/ClientFrontend.sol [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+import { Types } from "../libs/Types.sol";
+import { Transition } from "../libs/Transition.sol";
+import { Offchain } from "./Offchain.sol";
+
+contract ClientFrontend {
+    using SafeMath for uint256;
+    using Tx for bytes;
+    using Types for Types.UserState;
+
+    function decodeTransfer(bytes calldata txBytes)
```
Comment (vaibhavchellani) [inline contracts/client/ClientFrontend.sol:15]: all the variables named txBytes are encodedTxBytes correct? A small nitpick would be to change the variable name to reflect that, but you can ignore that if you don't feel a need for it

**Review by vaibhavchellani — APPROVED** [review]
LGTM!

File: contracts/client/ClientFrontend.sol (hunk 3 above)
Comment (ChihChengLiang) [reply contracts/client/ClientFrontend.sol:15]: Yes, they are encoded txBytes. I can rename those

**Pull request merged**